### PR TITLE
Add tightening_ratio support

### PR DESCRIPTION
**Pull request by nnmehta**
Add tightening_ratio support for tuning of the `TIGHENTING_RATIO` to allow scalable bloom objects to have a higher capacity at cost of higher false positive rate.

We allow the `TIGHTENING_RATIO` argument only if the command is replicated.

### CODE REVIEW

**Comment by KarthikSubbarao** [comment]
Can this PR be rebased? The digest change is merged in and looks like this PR contains commits from digest (which should be removed after rebasing)

File: src/bloom/command_handler.rs [hunk 1]
```diff
@@ -228,6 +230,16 @@ pub fn bloom_filter_reserve(ctx: &Context, input_args: &[ValkeyString]) -> Valke
...
+    let tightening_ratio = match input_args[curr_cmd_idx].to_string_lossy().parse::<f64>() {
+        Ok(num) if num > TIGHTENING_RATIO_MIN && num < TIGHTENING_RATIO_MAX => num,
+        Ok(num) if !(num > TIGHTENING_RATIO_MIN && num < TIGHTENING_RATIO_MAX) => {
+            return Err(ValkeyError::Str(utils::ERROR_RATIO_RANGE));
+        }
+        _ => {
+            return Err(ValkeyError::Str(utils::BAD_ERROR_RATIO));
+        }
+    };
+    curr_cmd_idx += 1;
```
Comment (KarthikSubbarao) [inline src/bloom/command_handler.rs]: This will not be provided as a command argument for the BF.RESERVE command. We can remove this

File: src/bloom/command_handler.rs [hunk 2]
```diff
@@ -327,6 +341,18 @@ pub fn bloom_filter_insert(ctx: &Context, input_args: &[ValkeyString]) -> Valkey
                     }
                 };
             }
+            "RATIO" => {
```
Comment (KarthikSubbarao) [inline src/bloom/command_handler.rs]: 1. Renaming to `TIGHTENING` makes this clearer
2. Let's add a check to only allow this argument if the command is replicated. If it is not replicated, we should not accept it as command args. This means, if the command is on a primary, it will use the constant from the configs.rs  file

File: src/bloom/command_handler.rs [hunk 3]
```diff
@@ -228,6 +230,16 @@ pub fn bloom_filter_reserve(ctx: &Context, input_args: &[ValkeyString]) -> Valke
         }
```
Comment (KarthikSubbarao) [inline src/bloom/command_handler.rs]: We will need to update the arity check at the beginning of the fn to handle the extra TIGHTENING argument

File: src/configs.rs [hunk 4]
```diff
@@ -32,6 +32,8 @@ lazy_static! {
 // Tightening ratio used during scale out for the calculation of fp_rate of every new filter within a bloom object to
 // maintain the bloom object's overall fp_rate to the configured value.
 pub const TIGHTENING_RATIO: f64 = 0.5;
+pub const TIGHTENING_RATIO_MIN: f64 = 0.001;
```
Comment (KarthikSubbarao) [inline src/configs.rs]: This should be 0, right?

File: src/wrapper/bloom_callback.rs [hunk 5]
```diff
@@ -2,6 +2,7 @@ use crate::bloom;
 use crate::bloom::data_type::ValkeyDataType;
 use crate::bloom::utils::BloomFilterType;
 use crate::configs;
+use crate::wrapper::digest::Digest;
```
Comment (KarthikSubbarao) [inline src/wrapper/bloom_callback.rs]: Since we are updating RDB Load, we need to update RDB Save to add the tightening ratio at the right place.

File: src/bloom/data_type.rs [hunk 6]
```diff
@@ -122,10 +127,28 @@ impl ValkeyDataType for BloomFilterType {
...
             filters,
         };
         Some(item)
     }
+
+    /// Function that is used to generate a digest on the Bloom Object.
+    fn debug_digest(&self, mut dig: Digest) {
+        dig.add_long_long(self.expansion.into());
+        dig.add_string_buffer(&self.fp_rate.to_le_bytes());
+        // dig.add_string_buffer(&self.tightening_ratio.to_le_bytes());
```
Comment (KarthikSubbarao) [inline src/bloom/data_type.rs]: We should add the tightening ratio to the digest. Why is this commented out?

File: src/bloom/command_handler.rs (hunk 1 above)
Comment (nnmehta) [reply src/bloom/command_handler.rs]: Sure will remove this

File: src/configs.rs (hunk 4 above)
Comment (nnmehta) [reply src/configs.rs]: Will update

File: src/wrapper/bloom_callback.rs (hunk 5 above)
Comment (nnmehta) [reply src/wrapper/bloom_callback.rs]: Will update [here](https://github.com/valkey-io/valkey-bloom/blob/7c254680374a22794a7676da5797a9f84b2e9087/src/wrapper/bloom_callback.rs#L23) in `rdb_save`

File: src/bloom/data_type.rs (hunk 6 above)
Comment (nnmehta) [reply src/bloom/data_type.rs]: I was testing out something else and commented this briefly, will uncomment in future revision

File: src/bloom/command_handler.rs (hunk 3 above)
Comment (nnmehta) [reply src/bloom/command_handler.rs]: Will update this

File: src/bloom/command_handler.rs [hunk 7]
```diff
@@ -299,14 +298,15 @@ pub fn bloom_filter_reserve(ctx: &Context, input_args: &[ValkeyString]) -> Valke
 pub fn bloom_filter_insert(ctx: &Context, input_args: &[ValkeyString]) -> ValkeyResult {
     let argc = input_args.len();
     // At the very least, we need: BF.INSERT <key> ITEMS <item>
-    if argc < 4 {
+    if argc < 5 {
```
Comment (KarthikSubbarao) [inline src/bloom/command_handler.rs]: This depends on `ctx.get_flags().contains(ContextFlags::REPLICATED)`, right?

If replica, < 5. If primary, < 4. Can we update it?

File: src/bloom/utils.rs [hunk 8]
```diff
@@ -94,9 +98,11 @@ impl BloomFilterType {
         // Create the bloom filter and add to the main BloomFilter object.
         let bloom = BloomFilter::new(fp_rate, capacity);
         let filters = vec![bloom];
+        let tightening_ratio = 0.5;
```
Comment (KarthikSubbarao) [inline src/bloom/utils.rs]: This will need to be passed in as arguments from `new_reserved`. This should not be hard coded

Comment (nnmehta) [reply src/bloom/utils.rs]: I had added it in this [change](https://github.com/nnmehta/ValkeyBloom/commit/cc18784ba7706352bb74c918128057feeb740bdd#diff-706cf6a8f58683c30caf89a9e9b6db3af4a4ed8df5c2af8cfe1bc8d6ce56d74aR118) but since you mentioned above: 
```
 not be provided as a command argument for the BF.RESERVE command
```

I had to remove it wherever we use `new_reserved`

File: src/bloom/command_handler.rs [hunk 9]
```diff
@@ -227,7 +227,6 @@ pub fn bloom_filter_reserve(ctx: &Context, input_args: &[ValkeyString]) -> Valke
             return Err(ValkeyError::Str(utils::BAD_ERROR_RATE));
         }
     };
-    curr_cmd_idx += 1;
```
Comment (KarthikSubbarao) [inline src/bloom/command_handler.rs]: Let's add this back. It might have been accidentally removed in your PR

File: src/bloom/utils.rs [hunk 10]
```diff
@@ -779,6 +794,7 @@ mod tests {
 
         // verify new_bf and bf
         assert_eq!(bf.fp_rate, new_bf.fp_rate);
+        assert_eq!(bf.tightening_ratio, new_bf.tightening_ratio);
```
Comment (KarthikSubbarao) [inline src/bloom/utils.rs:833]: In the verify_restored_items function, can we assert on this?
You can also do this assert for the false positive rate
```
        assert_eq!(
            restored_bloom_filter_type.tightenting_ratio,
            original_bloom_filter_type.tightenting_ratio
        );
```

File: src/bloom/command_handler.rs [hunk 11]
```diff
@@ -298,15 +302,17 @@ pub fn bloom_filter_reserve(ctx: &Context, input_args: &[ValkeyString]) -> Valke
 
 pub fn bloom_filter_insert(ctx: &Context, input_args: &[ValkeyString]) -> ValkeyResult {
     let argc = input_args.len();
+    let validate_size_limit = ctx.get_flags().contains(ContextFlags::REPLICATED);
```
Comment (KarthikSubbarao) [inline src/bloom/command_handler.rs]:
```suggestion
    let replicated_cmd = ctx.get_flags().contains(ContextFlags::REPLICATED);
```

File: src/bloom/command_handler.rs [hunk 12]
```diff
@@ -327,6 +333,21 @@ pub fn bloom_filter_insert(ctx: &Context, input_args: &[ValkeyString]) -> Valkey
                     }
                 };
             }
+            "TIGHTENING" => {
+                if !validate_size_limit {
+                    return Err(ValkeyError::Str(utils::ERROR));
+                }
```
Comment (KarthikSubbarao) [inline src/bloom/command_handler.rs]:
```suggestion
            "TIGHTENING" if replicated_cmd => {
                if idx >= (argc - 1) {
                    return Err(ValkeyError::WrongArity);
                }
```

File: src/bloom/command_handler.rs [hunk 13]
```diff
@@ -298,15 +302,17 @@ pub fn bloom_filter_reserve(ctx: &Context, input_args: &[ValkeyString]) -> Valke
 
 pub fn bloom_filter_insert(ctx: &Context, input_args: &[ValkeyString]) -> ValkeyResult {
     let argc = input_args.len();
+    let validate_size_limit = ctx.get_flags().contains(ContextFlags::REPLICATED);
     // At the very least, we need: BF.INSERT <key> ITEMS <item>
-    if argc < 4 {
+    if !validate_size_limit && argc < 4 || validate_size_limit && argc < 5 {
```
Comment (KarthikSubbarao) [inline src/bloom/command_handler.rs]:
```suggestion
    if argc < 4 {
```

This should handle it already. It is not required to contain tightening ratio on the replicated command - it is optional and we should handle cases where it provided and not provided.

The check above is sufficient

File: src/bloom/command_handler.rs [hunk 14]
```diff
@@ -405,6 +426,7 @@ pub fn bloom_filter_insert(ctx: &Context, input_args: &[ValkeyString]) -> Valkey
             }
             let mut bloom = match BloomFilterType::new_reserved(
                 fp_rate,
+                tightening_ratio,
```
Comment (KarthikSubbarao) [inline src/bloom/command_handler.rs:433]: Lets remove the following code at line 406
```
    // Skip bloom filter size validation on replicated cmds.
    let validate_size_limit = !ctx.get_flags().contains(ContextFlags::REPLICATED);
```

File: src/bloom/utils.rs [hunk 15]
```diff
@@ -95,9 +100,11 @@ impl BloomFilterType {
         // Create the bloom filter and add to the main BloomFilter object.
         let bloom = BloomFilter::new(fp_rate, capacity);
         let filters = vec![bloom];
+        let tightening_ratio = 0.5;
```
Comment (KarthikSubbarao) [inline src/bloom/utils.rs]: Why is this needed? Removing it
```suggestion
```

File: tests/test_replication.py [hunk 16]
```diff
@@ -34,7 +34,7 @@ def test_replication_behavior(self):
             ('BF.ADD', 'BF.ADD key item', 'BF.ADD key item1', 2),
             ('BF.MADD', 'BF.MADD key item', 'BF.MADD key item1', 2),
             ('BF.RESERVE', 'BF.RESERVE key 0.001 100000', 'BF.ADD key item1', 1),
-            ('BF.INSERT', 'BF.INSERT key items item', 'BF.INSERT key items item1', 2),
+            ('BF.INSERT', 'BF.INSERT key items item tightening 0.5', 'BF.INSERT key items item1 tightening 0.5', 2),
```
Comment (KarthikSubbarao) [inline tests/test_replication.py:37]: This is an invalid test. The tightening ratio is coming after `ITEM` so it is treated as an item. It is not treated as a property. Also, we cannot use this command on the primary, since we only allow this arg during replication :D 

Let's just use a normal BF.INSERT command.

Once we support the tightening ratio config, we can then (1) update the config on the primary node, (2) Use BF.INSERT (3) Validate Digest matches on the replica even though the tightening ratio config is different

Comment (KarthikSubbarao) [inline tests/test_replication.py]:
```suggestion
            ('BF.INSERT', 'BF.INSERT key items item', 'BF.INSERT key items item1', 2),
```

This is an invalid test. The tightening ratio is coming after `ITEM` so it is treated as an item. It is not treated as a property. Also, we cannot use this command on the primary, since we only allow this arg during replication :D 

Let's just use a normal BF.INSERT command.

Once we support the tightening ratio config, we can then (1) update the config on the primary node, (2) Use BF.INSERT (3) Validate Digest matches on the replica even though the tightening ratio config is different

For now, let us leave it as it is because it is essentially untested code (unless we use a monitor on replica during bloom creation on the primary node)

File: src/bloom/utils.rs [hunk 17]
```diff
@@ -642,12 +672,18 @@ mod tests {
         let rand_prefix = random_prefix(7);
         // 1 in every 1000 operations is expected to be a false positive.
         let expected_fp_rate: f64 = 0.001;
+        let expected_tightening_ratio: f64 = 0.5;
```
Comment (KarthikSubbarao) [inline src/bloom/utils.rs]:
```suggestion
        let tightening_ratio: f64 = 0.5;
```

File: src/bloom/utils.rs [hunk 18]
```diff
@@ -584,13 +608,19 @@ mod tests {
...
```
Comment (KarthikSubbarao) [inline src/bloom/utils.rs]:
```suggestion
        let tightening_ratio: f64 = 0.5;
```

We are not asserting against this. so there is no need to call it "expected"

File: src/bloom/utils.rs (hunk 15 above)
Comment (nnmehta) [reply src/bloom/utils.rs]: Missed to remove this, it is not required

File: tests/test_replication.py (hunk 16 above)
Comment (nnmehta) [reply tests/test_replication.py]: Ah got it, will revert this change accordingly

File: src/bloom/command_handler.rs (hunk 11 above)
Comment (nnmehta) [reply src/bloom/command_handler.rs]: Will update

File: src/bloom/utils.rs (hunk 18 above)
Comment (nnmehta) [reply src/bloom/utils.rs]: Will change accordingly

File: src/bloom/command_handler.rs (hunk 14 above)
Comment (nnmehta) [reply src/bloom/command_handler.rs:433]: Why would we want to remove this?
Can we discuss this?

**Review by KarthikSubbarao — APPROVED** [review]

File: src/bloom/utils.rs [hunk 19]
```diff
@@ -22,6 +25,8 @@ pub const BAD_EXPANSION: &str = "ERR bad expansion";
 pub const BAD_CAPACITY: &str = "ERR bad capacity";
 pub const BAD_ERROR_RATE: &str = "ERR bad error rate";
 pub const ERROR_RATE_RANGE: &str = "ERR (0 < error rate range < 1)";
+pub const BAD_ERROR_RATIO: &str = "ERR bad error ratio";
+pub const ERROR_RATIO_RANGE: &str = "ERR (0 < error ratio range < 1)";
```
Comment (zackcam) [inline src/bloom/utils.rs:29]: Would having a better error message i.e. something like `ratio must be in the range (0, 1)` or is following convention for error rate range better?

**Review by zackcam — APPROVED** [review]
